docs/roslyn_best_practices.md

```diff
@@ -0,0 +1,181 @@
+* Use information already available from `SyntaxNodeAnalysisContext` to quickly filter relevant information. As an example, if you build an analyzer that looks for a method then you would be evaluating `MemberAccessExpressionSyntax` nodes. In this scenario, you should also check the parent of the node to determine if this Syntax is a method or a property.
+
+**Do Not**
+* Do not put string compare operations above other conditional checks that can be performed. String comparisson is slower than checking for a type, or examining an enumeration value.
+
+
+### 2. Beware of String manipulations
+Build your analyzer with the expectation that it will be invoked a million times. Look for, and replace, strings to prevent excessive garbage collection due to frequent executions of your analyzer.
+
+**Do**
+* Use string constants when evaluating string conditionals.
+
+**Do Not**
+* Do not use string interpolation, contenation, or format strings which construct new objects each time the analyzer is run.
+
+### 3. Enable your Analyzer to run concurrently
+Enable concurrent execution of your analyzers and prevent them from running in auto generated code.
+
+**Do**
+* Use the following in your diagnostic analyzer class.
+```cs
+    public override void Initialize(AnalysisContext context)
+    {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            ...
+    }
+```
+
+### 4. Directly share node locations with the code fixer
+If you track multiple syntax nodes in your Analyzer then you can make it easier to write your Code Fixer by using the method overload of `Diagnostic.Create` that supports [additionalLocations](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostic.create#Microsoft_CodeAnalysis_Diagnostic_Create_Microsoft_CodeAnalysis_DiagnosticDescriptor_Microsoft_CodeAnalysis_Location_System_Collections_Generic_IEnumerable_Microsoft_CodeAnalysis_Location__System_Collections_Immutable_ImmutableDictionary_System_String_System_String__System_Object___)
+
+### 5. Beware of Trivia
+When searching for patterns of code it is common to perform string comparisons. These approaches can work successfully but you should beware that trivia, the spacing we use to make code more readable, can vary from team to team. Not everyone puts a space after an assignment operator and if there are two spaces, or a tab, after the assignment operator that code is still valid.
+
+**Should Not**
+* Try to avoid `ToFullString` and solutions that require awareness of trivia.
+
+### 6. Report the diagnostic only on parts of code that would be changed by a code fixer.
```

Comment:
What about focusing this practice as "Report the diagnostic only on part of the code that need to be changed". Whether or not that is manually or via a code fixer seems irrelevant.